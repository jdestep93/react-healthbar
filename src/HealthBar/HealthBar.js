import React from 'react';
import PropTypes from 'prop-types';
import {
  getHealthBarBackgroundColor,
  defaultColorPallet,
  hexColorRegex
} from './util/HealthBarUtils';

export default function HealthBar({
  percentage,
  colors = defaultColorPallet,
  width = 150,
  height = 24
}) {
  return (
    <div
      style={{
        borderRadius: '3px',
        border: '1px solid black',
        maxWidth: width,
        height
      }}
    >
      <div
        style={{
          width: `${percentage}%`,
          height: '100%',
          minHeight: height,
          backgroundColor: getHealthBarBackgroundColor(percentage, colors)
        }}
      >
        &nbsp;
      </div>
    </div>
  );
}

HealthBar.propTypes = {
  percentage: PropTypes.between({gte: 0, lte: 100}).isRequired,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  colors: PropTypes.arrayOf((propValue, key, componentName) => {
    if (!hexColorRegex.match(propValue[key])) {
      return new Error(
        `Invalid color code ${propValue[key]} supplied to ${componentName}`
      );
    }
  })
};
