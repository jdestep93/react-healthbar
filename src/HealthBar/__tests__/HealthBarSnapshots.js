import React from 'react';
import renderer from 'react-test-renderer';
import HealthBar from '../HealthBar'

describe('Health Bar Snapshots', () => {
  beforeEach(() => {
    // disable the console error so it doesnt appear in our tests
    jest.spyOn(global.console, 'error').mockImplementation(() => jest.fn());
  });

  it('should render correctly', () => {
    expect(
      renderer.create(
        <HealthBar
          percentage={100}
          width={200}
          height={30}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('should render correctly', () => {
    expect(
      renderer.create(
        <HealthBar
          percentage={0}
          width={100}
          height={10}
          colors={[
            '#9d1',
            '#ff1',
            '#c00'
          ]}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('should render correctly', () => {
    expect(
      renderer.create(
        <HealthBar
          percentage={55}
          width={100}
          height={200}
          colors={[
            '#9d1',
            '#ff1',
            '#c00',
            '#a00',
            '#b77',
            '#b42'
          ]}
        />
      ).toJSON()
    ).toMatchSnapshot();
  });

  it('should throw an error since theres no percentage', () => {
    expect(() =>
      renderer.create(
        <HealthBar
          width={100}
          height={200}
          colors={[
            '#9d1',
            '#ff1',
            '#c00',
            '#a00',
            '#b77',
            '#b42'
          ]}
        />
      )
    ).toThrow()
  });

  it('should throw if there is an invalid hex color', () => {
    try {
      renderer.create(
        <HealthBar
          percentage={32}
          width={100}
          height={200}
          colors={[
            '#9d1',
            '#ff1',
            '#c00',
            '#a00',
            '#b77',
            '#someinvalidhexcolor'
          ]}
        />
      )
    } catch (e) {
      expect(!!e).toBe(true);
    }
  });

  it('should throw if the percentage is below 0', () => {
    try {
      renderer.create(
        <HealthBar
          percentage={-1}
          width={100}
          height={200}
          colors={[
            '#9d1',
            '#ff1',
            '#c00',
            '#a00',
            '#b77',
            '#b42'
          ]}
        />
      )
    } catch (e) {
      expect(!!e).toBe(true);
    }
  });

  it('should throw if the percentage is above 100', () => {
    try {
      renderer.create(
        <HealthBar
          percentage={101}
          width={100}
          height={200}
          colors={[
            '#9d1',
            '#ff1',
            '#c00',
            '#a00',
            '#b77',
            '#b42'
          ]}
        />
      )
    } catch (e) {
      expect(!!e).toBe(true);
    }
  });
});
