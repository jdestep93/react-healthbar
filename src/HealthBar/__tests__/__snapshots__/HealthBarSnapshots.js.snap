// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Health Bar Snapshots should render correctly 1`] = `
<div
  style={
    Object {
      "border": "1px solid black",
      "borderRadius": "3px",
      "height": 30,
      "maxWidth": 200,
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "#15c621",
        "height": "100%",
        "minHeight": 30,
        "width": "100%",
      }
    }
  >
     
  </div>
</div>
`;

exports[`Health Bar Snapshots should render correctly 2`] = `
<div
  style={
    Object {
      "border": "1px solid black",
      "borderRadius": "3px",
      "height": 10,
      "maxWidth": 100,
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "#c00",
        "height": "100%",
        "minHeight": 10,
        "width": "0%",
      }
    }
  >
     
  </div>
</div>
`;

exports[`Health Bar Snapshots should render correctly 3`] = `
<div
  style={
    Object {
      "border": "1px solid black",
      "borderRadius": "3px",
      "height": 200,
      "maxWidth": 100,
    }
  }
>
  <div
    style={
      Object {
        "backgroundColor": "#c00",
        "height": "100%",
        "minHeight": 200,
        "width": "55%",
      }
    }
  >
     
  </div>
</div>
`;
